File: 2019/13/go_day13/day13.go
// Package main provides ...
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"strconv"
	"strings"
)

var (
	MODE_POSITION        = 0
	MODE_IMMEDIATE       = 1
	MODE_RELATIVE        = 2
	OP_ADD               = 1
	OP_MULT              = 2
	OP_SAVE              = 3
	OP_WRITE             = 4
	OP_JUMP_IF_TRUE      = 5
	OP_JUMP_IF_FALSE     = 6
	OP_LESS_THAN         = 7
	OP_EQUALS            = 8
	OP_SET_RELATIVE_BASE = 9
	OP_STOP              = 99
)

type Computer struct {
	Memory          []int
	Inputs          []int
	Outputs         []int
	PC              int
	RelativeBase    int
	Halted          bool
	WaitingForInput bool
}

func DigitFromRight(x, n int) int {
	tens := int(math.Pow(10, float64(n)))
	return x / tens % 10
}

// Solve takes a program and inputs, makes a computer, runs it and gets the output
func Solve(program []int, inputs []int) []int {
	c := NewComputer(program, inputs)
	c.Execute()
	return c.Outputs
}

// NewComputer initializes a new computer, given a program to run and inputs
func NewComputer(program []int, inputs []int) Computer {
	// memory := make([]int, len(program))
	// copy(memory, program)

	// Unsure if this approach works
	memory := make([]int, len(program)+10000)
	copy(memory, program)

	myInputs := make([]int, len(inputs))
	copy(myInputs, inputs)

	c := Computer{Memory: memory, Inputs: inputs, PC: 0, RelativeBase: 0}
	return c
}

// Direct gets the direct value of the memory address of the Nth arg, or PC + N
func (c Computer) Direct(n int) int {
	return c.Memory[c.PC+n]
}

// Lookup gets the dereferenced value of the Nth arg, after checking the Nth
// mode of the current instruction.
func (c Computer) Lookup(n int) int {
	instruction := c.Memory[c.PC]
	// If instruction is 105, and n=1, mode is the "1", or the 2nd digit
	// from right 0 indexed (3rd when counting naturally)
	mode := DigitFromRight(instruction, n+1)
	if mode == MODE_POSITION {
		return c.Memory[c.Direct(n)]
	} else if mode == MODE_IMMEDIATE {
		return c.Direct(n)
	} else if mode == MODE_RELATIVE {
		address := c.Direct(n) + c.RelativeBase
		return c.Memory[address]
	} else {
		log.Fatal("Unknown mode")
	}
	return 0
}

func (c Computer) LookupLeft(n int) int {
	// Now: LookupLeft is same as direct
	instruction := c.Memory[c.PC]
	mode := DigitFromRight(instruction, n+1)
	if mode == MODE_POSITION {
		return c.Direct(n)
	} else if mode == MODE_IMMEDIATE {
		return c.Direct(n)
	} else if mode == MODE_RELATIVE {
		address := c.Direct(n) + c.RelativeBase
		return address
	} else {
		log.Fatal("Unknown mode")
	}
	return 0
}

// AddInput adds an input to a computer
func (c *Computer) AddInput(newInput int) {
	c.Inputs = append(c.Inputs, newInput)
}

// PopOutput returns the oldest output, and removes it from the computer's output
func (c *Computer) PopOutput() (int, error) {
	if len(c.Outputs) == 0 {
		return 0, errors.New("Computer has no output, can't pop")
	}
	first, rest := c.Outputs[0], c.Outputs[1:]
	c.Outputs = rest
	return first, nil
}

// Execute runs instructions until the computer stops (99) or pauses (Waiting for input)
func (c *Computer) Execute() {
	// Stopped: c.Halted == true || (c.WaitingForInput && len(c.Inputs) == 0)
	// Running is inverse
	for c.Halted == false && (!c.WaitingForInput || len(c.Inputs) > 0) {
		c.Step()
	}
}

// Step executes the next instruction and moves the PC
func (c *Computer) Step() {
	instruction := c.Memory[c.PC] % 100
	if instruction == OP_ADD {
		c.Memory[c.LookupLeft(3)] = c.Lookup(1) + c.Lookup(2)
		c.PC += 4
	} else if instruction == OP_MULT {
		c.Memory[c.LookupLeft(3)] = c.Lookup(1) * c.Lookup(2)
		c.PC += 4
	} else if instruction == OP_SAVE {
		if len(c.Inputs) == 0 {
			// No input available, pause execution
			c.WaitingForInput = true
		} else {
			// Process Input
			c.WaitingForInput = false
			input := c.Inputs[0]
			c.Inputs = c.Inputs[1:]
			c.Memory[c.LookupLeft(1)] = input
			c.PC += 2
		}
	} else if instruction == OP_WRITE {
		c.Outputs = append(c.Outputs, c.Lookup(1))
		c.PC += 2
	} else if instruction == OP_JUMP_IF_TRUE {
		if c.Lookup(1) != 0 {
			c.PC = c.Lookup(2)
		} else {
			c.PC += 3
		}
	} else if instruction == OP_JUMP_IF_FALSE {
		if c.Lookup(1) == 0 {
			c.PC = c.Lookup(2)
		} else {
			c.PC += 3
		}
	} else if instruction == OP_LESS_THAN {
		if c.Lookup(1) < c.Lookup(2) {
			c.Memory[c.LookupLeft(3)] = 1
		} else {
			c.Memory[c.LookupLeft(3)] = 0
		}
		c.PC += 4
	} else if instruction == OP_EQUALS {
		if c.Lookup(1) == c.Lookup(2) {
			c.Memory[c.LookupLeft(3)] = 1
		} else {
			c.Memory[c.LookupLeft(3)] = 0
		}
		c.PC += 4
	} else if instruction == OP_SET_RELATIVE_BASE {
		c.RelativeBase += c.Lookup(1)
		c.PC += 2
	} else if instruction == OP_STOP {
		c.Halted = true
	} else {
		log.Fatalf("Unknown instruction %v", instruction)
	}
}

// AmplifyOnceMaxSeq takes a program given, and runs
// AmplifyOnce with many different 'phase sequence' values,
// (all permutations of (01234)).  It finds the sequence
// that returns the highest value, then returns both the
// sequence and the value.
func AmplifyOnceMaxSeq(program []int) ([]int, int) {
	maxValue := 0
	maxSeq := []int{}

	for _, seq := range permutations([]int{0, 1, 2, 3, 4}) {
		value := AmplifyOnce(program, seq)
		if value > maxValue {
			maxValue = value
			maxSeq = seq
		}
	}
	return maxSeq, maxValue
}

// AmplifyOnce sets up a chain of 5 computers, loads them with a program,
// initializes them with the "Phase Sequence" values, sends a 0 to the first
// computer, sends the first computer's output to the second computer, etc,
// then returns the last computer's output.
func AmplifyOnce(program []int, phaseSeq []int) int {
	computers := make([]*Computer, 0)
	for _, num := range phaseSeq {
		c := NewComputer(program, []int{num})
		c.Execute()
		computers = append(computers, &c)
	}

	lastOutput := 0
	var err error
	for _, c := range computers {
		c.AddInput(lastOutput)
		c.Execute()
		lastOutput, err = c.PopOutput()
		if err != nil {
			log.Fatal("AmplifyOnce: Couldn't read output")
		}
	}
	return lastOutput
}

// AmplifyLoopMaxSeq takes a program given, and runs
// AmplifyLoop with many different 'phase sequence' values,
// (all permutations of (56789)).  It finds the sequence
// that returns the highest value, then returns both the
// sequence and the value.
func AmplifyLoopMaxSeq(program []int) ([]int, int) {
	maxValue := 0
	maxSeq := []int{}

	for _, seq := range permutations([]int{5, 6, 7, 8, 9}) {
		value := AmplifyLoop(program, seq)
		if value > maxValue {
			maxValue = value
			maxSeq = seq
		}
	}
	return maxSeq, maxValue
}

// AmplifyLoop sets up a chain of 5 computers, loads them with a program,
// initializes them with the "Phase Sequence" values, sends a 0 to the first
// computer, sends the first computer's output to the second computer, etc.
// The last computer's output is fed to the first computer's input in a feedback
// loop.  The signal keeps looping until at least one of the computers has halted,
// then it finishes the current cycle and returns the last computer's output.
func AmplifyLoop(program []int, phaseSeq []int) int {
	computers := make([]*Computer, 0)
	for _, num := range phaseSeq {
		c := NewComputer(program, []int{num})
		c.Execute()
		computers = append(computers, &c)
	}

	lastOutput := 0
	var err error
	stillRunning := true

	for stillRunning && lastOutput >= 0 {
		for _, c := range computers {
			c.AddInput(lastOutput)
			c.Execute()
			lastOutput, err = c.PopOutput()
			if err != nil {
				log.Fatal("AmplifyLoop: Couldn't read output")
			}
			if c.Halted {
				stillRunning = false
			}
		}
	}
	return lastOutput
}

func Parse(filename string) []int {
	program := make([]int, 0)

	b, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}
	s := string(b)
	sList := strings.Split(strings.TrimSpace(s), ",")
	for _, sNum := range sList {
		num, err := strconv.Atoi(sNum)
		if err != nil {
			log.Fatal(err)
		}
		program = append(program, num)
	}
	return program
}

// https://stackoverflow.com/questions/30226438/generate-all-permutations-in-go
func permutations(arr []int) [][]int {
	var helper func([]int, int)
	res := [][]int{}

	helper = func(arr []int, n int) {
		if n == 1 {
			tmp := make([]int, len(arr))
			copy(tmp, arr)
			res = append(res, tmp)
		} else {
			for i := 0; i < n; i++ {
				helper(arr, n-1)
				if n%2 == 1 {
					tmp := arr[i]
					arr[i] = arr[n-1]
					arr[n-1] = tmp
				} else {
					tmp := arr[0]
					arr[0] = arr[n-1]
					arr[n-1] = tmp
				}
			}
		}
	}
	helper(arr, len(arr))
	return res
}

///////// PREVIOUS INTCODE END //////////////

type Coord struct {
	X, Y int
}

// Move moves a coordinate by one step in the specified direction.
func Move(where Coord, direction string) Coord {
	switch direction {
	case "R":
		return Coord{where.X + 1, where.Y}
	case "L":
		return Coord{where.X - 1, where.Y}
	case "U":
		return Coord{where.X, where.Y - 1}
	case "D":
		return Coord{where.X, where.Y + 1}
	}
	log.Fatal("Move: Unknown direction")
	return where
}

func TurnRight(direction string) string {
	switch direction {
	case "U":
		return "R"
	case "R":
		return "D"
	case "D":
		return "L"
	case "L":
		return "U"
	}
	log.Fatal("TurnRight: Unknown direction")
	return "Z"
}

func TurnLeft(direction string) string {
	switch direction {
	case "U":
		return "L"
	case "L":
		return "D"
	case "D":
		return "R"
	case "R":
		return "U"
	}
	log.Fatal("TurnLeft: Unknown direction")
	return "Z"
}

func PainterRobot(program []int, initialColor int) map[Coord]int {
	c := NewComputer(program, []int{})
	grid := make(map[Coord]int, 0)

	location := Coord{0, 0}
	direction := "U"
	grid[location] = initialColor

	for {
		currentSquare := grid[location]
		c.AddInput(currentSquare)
		c.Execute()
		if c.Halted {
			break
		}
		paintColor, err := c.PopOutput()
		if err != nil {
			log.Fatal("PainterRobot: Couldn't read paintColor")
		}
		turnDir, err := c.PopOutput()
		if err != nil {
			log.Fatal("PainterRobot: Couldn't read turnDir")
		}
		grid[location] = paintColor
		if turnDir == 1 {
			direction = TurnRight(direction)
		} else if turnDir == 0 {
			direction = TurnLeft(direction)
		} else {
			log.Fatal("Told to turn an incorrect direction")
		}
		location = Move(location, direction)
	}

	return grid
}

func NumPaintedSquares(grid map[Coord]int) int {
	i := 0
	for range grid {
		i++
	}
	return i
}

func DisplayGrid(grid map[Coord]int) {
	// for y in range(-3, 7):
	//     for x in range(-5, 45):
	for y := -1; y < 7; y++ {
		for x := -5; x < 45; x++ {
			value := grid[Coord{x, y}]
			if value == 1 {
				fmt.Printf("#")
			} else {
				fmt.Printf(" ")
			}
		}
		fmt.Println("")
	}
}

func main() {
	fmt.Println("Advent of Code 2019 Day 9")

	program := Parse("../../11/input.txt")
	grid1 := PainterRobot(program, 0)
	fmt.Println("Part 1:")
	fmt.Println(NumPaintedSquares(grid1))

	grid2 := PainterRobot(program, 1)
	fmt.Println("Part 2:")
	fmt.Println(NumPaintedSquares(grid2))
	DisplayGrid(grid2)
}

File: 2019/13/python_day13/day13.py
#!/usr/bin/env python

from itertools import permutations
from collections import defaultdict
from os import system
import sys

# DEBUG = True
DEBUG = False


class OP:
    ADD = 1
    MULT = 2
    SAVE = 3
    WRITE = 4
    JUMP_IF_TRUE = 5
    JUMP_IF_FALSE = 6
    LESS_THAN = 7
    EQUALS = 8
    SET_REL_BASE = 9
    STOP = 99


class MODE:
    POSITION = 0
    IMMEDIATE = 1
    RELATIVE = 2


def digit_from_right(x, n):
    return x // (10 ** n) % 10


class Computer(object):
    def __init__(self, memory, inputs):
        self.memory = defaultdict(lambda: 0)
        for i, k in enumerate(memory):
            self.memory[i] = k
        self.inputs = inputs.copy()
        self.outputs = []
        self.pc = 0
        self.state = "new"
        self.relative_base = 0

    def direct(self, n):
        """ Get the direct value of the memory address of the Nth arg, or PC + N"""
        return self.memory[self.pc + n]

    def lookup(self, n):
        """ Get the dereferenced value of the Nth arg, after checking the Nth mode
        of the current instruction. """
        instruction = self.memory[self.pc]
        # If instruction is 105, and n=1, mode is the "1", or the 2nd digit
        # from right 0 indexed (3rd when counting naturally)
        mode = digit_from_right(instruction, n + 1)
        if mode == MODE.POSITION:
            return self.memory[self.direct(n)]
        if mode == MODE.IMMEDIATE:
            return self.direct(n)
        if mode == MODE.RELATIVE:
            # Like position, but counts from relative base
            address = self.direct(n) + self.relative_base
            return self.memory[address]
        raise Exception("Unknown mode")

    def lookup_left(self, n):
        """ Use on left side of equals only (needs better explanation) """
        instruction = self.memory[self.pc]
        mode = digit_from_right(instruction, n + 1)
        if mode == MODE.POSITION:
            return self.direct(n)
        if mode == MODE.IMMEDIATE:
            return self.direct(n)
        if mode == MODE.RELATIVE:
            return self.direct(n) + self.relative_base

    def info(self, string):
        if DEBUG:
            print(string)

    def add_input(self, x):
        self.inputs.append(x)

    def has_output(self):
        return len(self.outputs) > 0

    def pop_output(self):
        return self.outputs.pop(0)

    def execute(self):
        if self.state == "halted":
            print("Refusing to execute; is halted")
        self.state = "running"
        while True:
            instruction = self.memory[self.pc] % 100
            # print(self.memory[self.pc])
            if instruction == OP.ADD:
                self.memory[self.lookup_left(3)] = self.lookup(1) + self.lookup(2)
                self.info(
                    f"    -> ADD program[{self.direct(3)}] = {self.lookup(1)} + {self.lookup(2)} = {self.lookup(1) + self.lookup(2)}"
                )
                self.pc += 4
            elif instruction == OP.MULT:
                self.memory[self.lookup_left(3)] = self.lookup(1) * self.lookup(2)
                self.info(
                    f"    -> ADD program[{self.direct(3)}] = {self.lookup(1)} * {self.lookup(2)} = {self.lookup(1) * self.lookup(2)}"
                )
                self.pc += 4
            elif instruction == OP.SAVE:
                if len(self.inputs) == 0:
                    # print("Not enough input!")
                    self.state = "waiting_input"
                    break
                this_input = self.inputs.pop(0)

                self.memory[self.lookup_left(1)] = this_input

                self.info(
                    f"    -> SAVE program[{self.lookup(1)}] = INPUT = {this_input}"
                )
                self.pc += 2
            elif instruction == OP.WRITE:
                self.outputs.append(self.lookup(1))
                self.info(f"    -> WRITE {self.lookup(1)} = OUTPUT")
                self.pc += 2
            elif instruction == OP.JUMP_IF_TRUE:
                if self.lookup(1) != 0:
                    self.info(
                        f"    -> JUMP_IF_TRUE [{self.lookup(1)}] != 0, setting i = [{self.lookup(2)}]"
                    )
                    self.pc = self.lookup(2)
                else:
                    self.info(
                        f"    -> JUMP_IF_TRUE [{self.lookup(1)}] == 0, doing normal i+= 3"
                    )
                    self.pc += 3
            elif instruction == OP.JUMP_IF_FALSE:
                if self.lookup(1) == 0:
                    self.info(
                        f"    -> JUMP_IF_FALSE [{self.lookup(1)}] == 0, setting i = [{self.lookup(2)}]"
                    )
                    self.pc = self.lookup(2)
                else:
                    self.info(
                        f"    -> JUMP_IF_FALSE [{self.lookup(1)}] != 0, doing normal i+= 3"
                    )
                    self.pc += 3
            elif instruction == OP.LESS_THAN:
                if self.lookup(1) < self.lookup(2):
                    self.info(
                        f"    -> LESS_THAN [{self.lookup(1)}] < [{self.lookup(2)}], setting program[{self.lookup_left(3)}] = 1"
                    )
                    self.memory[self.lookup_left(3)] = 1
                else:
                    self.info(
                        f"    -> LESS_THAN [{self.lookup(1)}] not < [{self.lookup(2)}], setting program[{self.lookup_left(3)}] = 0"
                    )
                    self.memory[self.lookup_left(3)] = 0
                self.pc += 4
            elif instruction == OP.EQUALS:
                if self.lookup(1) == self.lookup(2):
                    self.info(
                        f"    -> EQUALS [{self.lookup(1)}] == [{self.lookup(2)}], setting program[{self.lookup_left(3)}] = 1"
                    )
                    self.memory[self.lookup_left(3)] = 1
                else:
                    self.info(
                        f"    -> EQUALS [{self.lookup(1)}] != [{self.lookup(2)}], setting program[{self.lookup_left(3)}] = 0"
                    )
                    self.memory[self.lookup_left(3)] = 0
                self.pc += 4
            elif instruction == OP.SET_REL_BASE:
                adj = self.lookup(1)
                self.relative_base += adj
                self.info(
                    f"    -> ADJUST RELATIVE BASE by [{adj}]. NEW BASE = [{self.relative_base}]"
                )
                self.pc += 2
            elif instruction == OP.STOP:
                self.state = "halted"
                break


def parse(filename):
    with open(filename) as f:
        return [int(num) for num in f.readline().strip().split(",")]


def solve1(program_in, inputs):
    """ Given a program and inputs, make a new VM, run the program, and return
    its outputs when it stops. """
    c = Computer(program_in, inputs)
    c.execute()
    return c.outputs


def amplify_once_find_max_seq(program_in):
    """Try every combination of phase settings on the amplifiers. What is the
    highest signal that can be sent to the thrusters? (Max Val)"""
    max_val = 0
    max_sequence = []

    for seq in permutations([0, 1, 2, 3, 4]):
        phase_sequence = list(seq)
        val = amplify_once(program_in, phase_sequence)
        if val > max_val:
            max_val = val
            max_sequence = phase_sequence
    return [max_val, max_sequence]


def amplify_once(program_in, phase_sequence):
    input_signal = 0
    for setting in phase_sequence:
        outputs = solve1(program_in, [setting, input_signal])
        input_signal = outputs[0]
    return input_signal


def amplify_loop(program_in, phase_sequence):
    cpus = []
    for i in range(5):
        cpus.append(Computer(program_in, [phase_sequence[i]]))

    i = 0
    next_input = 0
    while True:
        cpus[i].add_input(next_input)
        cpus[i].execute()
        if cpus[i].state == "halted" and i == 4:
            # print("halted")
            return cpus[i].outputs[0]
        next_input = cpus[i].pop_output()
        i = (i + 1) % 5


def amplify_loop_max_seq(program_in):
    max_val = 0
    max_sequence = []
    for seq in permutations([5, 6, 7, 8, 9]):
        phase_sequence = list(seq)
        val = amplify_loop(program_in, phase_sequence)
        if val > max_val:
            max_val = val
            max_sequence = phase_sequence
    return [max_val, max_sequence]


COMPLEX_OF_DIR = {
    "U": complex(0, -1),
    "R": complex(1, 0),
    "D": complex(0, 1),
    "L": complex(-1, 0),
}


def turn_right(direction):
    return turn(direction, 1)


def turn_left(direction):
    return turn(direction, -1)


def turn(direction, n):
    directions = list(COMPLEX_OF_DIR.keys())
    new_index = (directions.index(direction) + n) % len(directions)
    return directions[new_index]


class PainterRobot(object):
    def __init__(self, program):
        self.program = program
        self.cpu = Computer(self.program, [])
        self.grid = defaultdict(lambda: 0)

    def input(self, val):
        self.cpu.add_input(val)

    def add_quarters(self):
        self.cpu.memory[0] = 2

    def run_and_return_grid(self):
        cpu = self.cpu
        self.cpu.memory[0] = 2
        grid = self.grid

        cpu.execute()
        while cpu.has_output():
            x = cpu.pop_output()
            y = cpu.pop_output()
            what = cpu.pop_output()
            grid[complex(x, y)] = what

        return grid

    def part2(self):
        return 5


def get_move(grid):
    ball = list(grid.keys())[list(grid.values()).index(4)]
    ball_x = int(ball.real)
    paddle = list(grid.keys())[list(grid.values()).index(3)]
    paddle_x = int(paddle.real)
    if ball_x > paddle_x:
        return "r"
    if ball_x < paddle_x:
        return "l"
    return "n"


def part1(program_in):
    robot = PainterRobot(program_in)
    grid = robot.run_and_return_grid()
    robot.add_quarters()
    display(grid)
    while True:
        a = get_move(grid)
        # a = input()
        if a == "l":
            robot.input(-1)
        elif a == "r":
            robot.input(1)
        else:
            robot.input(0)
        grid = robot.run_and_return_grid()
        display(grid)

    # print(grid)
    # print(list(grid.values()).count(2))
    return len(list(grid.keys()))


def display(grid):
    system("clear")
    for y in range(25):
        for x in range(45):
            char = grid[complex(x, y)]
            print_char = " "
            if char == 1:
                print_char = "W"
            elif char == 2:
                print_char = "B"
            elif char == 3:
                print_char = "="
            elif char == 4:
                print_char = "*"
            print(print_char, end="")
        print("")
    print(f"Score {grid[complex(-1, 0)]}")


def part2(program_in):
    robot = PainterRobot(program_in)
    grid = robot.run_and_return_grid()
    for y in range(-3, 7):
        for x in range(-5, 45):
            value = grid[complex(x, y)]
            if value == 1:
                print("#", end="")
            else:
                print(" ", end="")
        print("")
    return "Look above ^"


if __name__ == "__main__":
    system("clear")
    a = input()
    program = parse("../../13/input.txt")
    print(part1(program))
    # print("Part 2:")
    # print(part2(program))
